### PR TITLE
[HINTS] Add implementation for basic hints on dict.cairo + default_dict.cairo

### PR DESCRIPTION
**Pull request by fmoletta**
HINT STATUS

`dict_new`
- [x] Implementation
- [x] Unit Tests
- [x] Integration Bad case

`dict_read`
- [x] Implementation
- [x] Unit Tests
- [x] Integration Test  Good Case

`dict_write`
- [x] Implementation
- [x] Unit Tests
- [x] Integration Test  Good Case

`default_dict_new`
- [x] Implementation
- [x] Unit Tests
- [x] Integration Test  Good Case

dict.cairo hints not contemplated in this PR: `dict_update`, `dict_squash`(2)

Added features:
- Implementation of above hints on dict_hint_utils.rs
- Add Structures used to represent and manager dictionaries: DicManager, DictTracker, Dictionary on dict_manager.rs
- Add Dictionary variant to PyValueTypes in exec_scopes.rs
- Add bad_cairo_programs folder to store cairo programs that produce run-time errors for testing purposes
- Add compiling and cleanup of files in bad_cairo_programs to Makefile

### CODE REVIEW

File: src/vm/errors/vm_errors.rs [hunk 1]
```diff
@@ -150,9 +155,22 @@ impl fmt::Display for VirtualMachineError {
             VirtualMachineError::AssertNotZero(value, prime) => {
                 write!(f, "Assertion failed, {} % {} is equal to 0", value, prime)
             },
+            VirtualMachineError::CantCreateDictionaryOnTakenSegment(index) => {
+                write!(f, "Error: Tried to create a dictionary on segment: {:?} when this segment already corresponds to a dictionary", index)
```
Comment (Oppen) [inline src/vm/errors/vm_errors.rs:159]: Can't the segment belong to anything else? Is it being taken a guarantee that a dict took it? Essentially, could the message be misleading?

File: src/vm/hints/dict_hint_utils.rs [hunk 2]
```diff
@@ -0,0 +1,1239 @@
...
+
+use super::{dict_manager::DictManager, hint_utils::get_address_from_reference};
+//DictAccess struct has three memebers, so the size of DictAccess* is 3
+const DICT_ACCESS_SIZE: usize = 3;
+
+fn get_initial_dict(vm: &mut VirtualMachine) -> Option<HashMap<BigInt, BigInt>> {
+    let mut initial_dict: Option<HashMap<BigInt, BigInt>> = None;
+    if let Some(variables) = vm.exec_scopes.get_local_variables() {
+        if let Some(PyValueType::Dictionary(py_initial_dict)) =
+            variables.get(&String::from("initial_dict"))
```
Comment (Oppen) [inline src/vm/hints/dict_hint_utils.rs:18]:
```suggestion
            variables.get("initial_dict")
```

File: src/vm/hints/dict_hint_utils.rs [hunk 3]
```diff
@@ -0,0 +1,1239 @@
...
+    if vm.dict_manager.is_none() {
+        vm.dict_manager = Some(DictManager::new());
+    }
+
+    //Get initial dictionary from scope (defined by an earlier hint)
+    let initial_dict = if let Some(initial_dict) = get_initial_dict(vm) {
+        initial_dict
+    } else {
+        return Err(VirtualMachineError::NoInitialDict);
+    };
```
Comment (Oppen) [inline src/vm/hints/dict_hint_utils.rs:48]:
```suggestion
    let initial_dict = get_initial_dict(vm).ok_or(VirtualMachineError::NoInitialDict)?;
```

File: src/vm/vm_core.rs [hunk 4]
```diff
@@ -34,11 +35,14 @@ pub struct HintData {
...
     pub ids: HashMap<String, BigInt>,
 }
+
 pub struct VirtualMachine {
     pub run_context: RunContext,
     pub prime: BigInt,
     pub builtin_runners: Vec<(String, Box<dyn BuiltinRunner>)>,
     pub segments: MemorySegmentManager,
+    pub _program_base: Option<MaybeRelocatable>,
+    pub memory: Memory,
```
Comment (Oppen) [inline src/vm/vm_core.rs:45]: Why did these moved?

File: src/vm/errors/vm_errors.rs (hunk 1 above)
Comment (fmoletta) [reply src/vm/errors/vm_errors.rs:159]: Yes, this error is only produced by the DictManager when adding a tracker for the segment, so it being taken can only mean a dictionary was created on that segment. As it is an internal error (it shouldn't reach the user), I can make the error message a bit more clear, by making it more specific.

File: src/vm/hints/dict_hint_utils.rs (hunk 2 above)
Comment (fmoletta) [reply src/vm/hints/dict_hint_utils.rs]: Included!

File: src/vm/hints/dict_hint_utils.rs (hunk 3 above)
Comment (fmoletta) [reply src/vm/hints/dict_hint_utils.rs]: Included!

File: src/vm/vm_core.rs [hunk 5]
```diff
@@ -50,15 +54,14 @@ pub struct VirtualMachine {
...
-    pub _program_base: Option<MaybeRelocatable>,
-    pub memory: Memory,
     //auto_deduction: HashMap<BigInt, Vec<(Rule, ())>>,
     //Some(accessed_addresses) == proof mode enabled
     accessed_addresses: Option<Vec<MaybeRelocatable>>,
     //None if trace is not enabled, Some otherwise
     pub trace: Option<Vec<TraceEntry>>,
     current_step: usize,
     skip_instruction_execution: bool,
+    pub dict_manager: Option<DictManager>,
```
[the remark this answers is not in the document]
Comment (fmoletta) [reply src/vm/vm_core.rs]: Implemented change

**Review by Oppen — APPROVED** [review]

**Pull request merged**